File: examples/typescript/phone_call_burst.ts
// N concurrent live phone calls → DetectDeepfake.
//
// The recommended multi-call integration pattern: when a single PBX (e.g.
// a FreeSWITCH instance running `mod_audio_fork`) is handling many calls
// simultaneously, you want **one** long-lived gRPC channel multiplexing
// N concurrent bidi streams — not N separate channels. This file shows
// exactly that shape:
//
//   - One `new DeepfakeDetectionClient(target, ...)` for the whole run.
//   - N concurrent `client.detectDeepfake()` bidi streams over that channel,
//     each running the same `sendCall` / `recvCall` loop that
//     `phone_call.ts` documents in detail (imported, not duplicated).
//   - Per-stream `call-NN` label prefixes every log line so interleaved
//     output is grep-friendly.
//   - Optional `--stagger-ms` to spread call starts out (mimics a real
//     PBX where calls arrive over time, not all at exactly t=0).
//   - Graceful SIGINT/SIGTERM via `common.installSignalShutdown` cancels
//     every in-flight stream cleanly — no orphan bidi sockets, summary
//     still prints.
//   - Optional `--csv PATH` captures per-chunk results across all streams
//     in one file — useful for finding the connection-count knee on a
//     real backend.
//
// CLI:
//   tsx phone_call_burst.ts [--audio FILE] [--target localhost:50051]
//                           [--chunk-ms 100] [--duration 30]
//                           [-c|--concurrency 5] [--stagger-ms 500]
//                           [--scenario-id ID] [--csv PATH]

import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { Metadata, type ServiceError } from "@grpc/grpc-js";
import { DeepfakeDetectionClient } from "@aurigin/protos/aurigin/deepfake_detection/v1/deepfake_detection";

import {
  ResultCSV,
  type WavData,
  channelCredentials,
  durationS as wavDurationS,
  installSignalShutdown,
  readWav,
  transportLabel,
} from "./common/index.js";
// The per-call building blocks. `phone_call_burst` is "sendCall + recvCall
// instantiated N times over one channel" — importing keeps that
// relationship explicit and prevents drift between the two files' loops.
import { type Call, type ResponseSink, recvCall, sendCall } from "./phone_call.js";

const DEFAULT_CHUNK_MS = 100;
const DEFAULT_DURATION_S = 30;
const DEFAULT_CONCURRENCY = 1;

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

function resolveAudio(arg: string | null): string {
  if (arg) {
    if (!fs.existsSync(arg)) throw new Error(`No such file: ${arg}`);
    return arg;
  }
  const audioDir = path.join(__dirname, "..", "audio");
  if (!fs.existsSync(audioDir)) {
    throw new Error("No --audio supplied and examples/audio/ does not exist.");
  }
  const wavs = fs.readdirSync(audioDir).filter((f) => f.endsWith(".wav")).sort();
  if (wavs.length === 0) {
    throw new Error(
      "No --audio supplied and no .wav files found in examples/audio/. " +
        "Drop a 16-bit PCM or 32-bit IEEE-float WAV in examples/audio/ or pass one with --audio.",
    );
  }
  return path.join(audioDir, wavs[0]);
}

function makeLabels(n: number): string[] {
  const width = Math.max(2, String(n).length);
  return Array.from({ length: n }, (_, i) => `call-${String(i + 1).padStart(width, "0")}`);
}

interface CallResult {
  label: string;
  error: unknown | null;
}

async function runOneCall(
  client: DeepfakeDetectionClient,
  label: string,
  wav: WavData,
  chunkMs: number,
  durationS: number,
  metadata: Metadata | undefined,
  delayMs: number,
  csv: ResultCSV | null,
  fileName: string,
  registerCall: (call: Call) => void,
): Promise<CallResult> {
  // Stagger-driven delay before opening the stream — used by --stagger-ms
  // to spread call starts across wallclock.
  if (delayMs > 0) await sleep(delayMs);
  const call: Call = metadata ? client.detectDeepfake(metadata) : client.detectDeepfake();
  registerCall(call);
  const sink: ResponseSink = {
    sessionId: "", chunks: [], audioDurationMs: 0, globalResult: "unknown",
  };
  const tStart = performance.now();
  try {
    await Promise.all([
      sendCall(call, wav, chunkMs, durationS),
      recvCall(call, { label, sink }),
    ]);
    return { label, error: null };
  } catch (err) {
    return { label, error: err };
  } finally {
    if (csv) {
      csv.writeSession(
        fileName, sink.sessionId, sink.chunks,
        sink.audioDurationMs, sink.globalResult,
        performance.now() - tStart,
      );
    }
  }
}

// ─── argv parser ────────────────────────────────────────────────────────────

interface Args {
  audio: string | null;
  duration: number;
  chunkMs: number;
  target: string;
  concurrency: number;
  staggerMs: number;
  scenarioId: string | null;
  csv: string | null;
}

function parseArgs(argv: string[]): Args {
  const out: Args = {
    audio: null,
    duration: DEFAULT_DURATION_S,
    chunkMs: DEFAULT_CHUNK_MS,
    target: "localhost:50051",
    concurrency: DEFAULT_CONCURRENCY,
    staggerMs: 0,
    scenarioId: null,
    csv: null,
  };
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    const next = () => argv[++i];
    switch (arg) {
      case "--audio": out.audio = next(); break;
      case "--duration": out.duration = Number(next()); break;
      case "--chunk-ms": out.chunkMs = Number(next()); break;
      case "--target": out.target = next(); break;
      case "-c": case "--concurrency": out.concurrency = Number(next()); break;
      case "--stagger-ms": out.staggerMs = Number(next()); break;
      case "--scenario-id": out.scenarioId = next(); break;
      case "--csv": out.csv = next(); break;
      case "-h": case "--help":
        console.log("Usage: tsx phone_call_burst.ts [--audio FILE] [--duration SEC]");
        console.log("                               [--chunk-ms MS] [--target HOST:PORT]");
        console.log("                               [-c|--concurrency N] [--stagger-ms MS]");
        console.log("                               [--scenario-id ID] [--csv PATH]");
        process.exit(0);
      default: throw new Error(`Unknown arg: ${arg}`);
    }
  }
  if (!Number.isInteger(out.concurrency) || out.concurrency < 1) {
    throw new Error("--concurrency must be a positive integer");
  }
  if (!Number.isFinite(out.staggerMs) || out.staggerMs < 0) {
    throw new Error("--stagger-ms must be >= 0");
  }
  return out;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const audioPath = resolveAudio(args.audio);
  const wav = readWav(audioPath);
  const labels = makeLabels(args.concurrency);

  const metadata = args.scenarioId ? new Metadata() : undefined;
  if (metadata && args.scenarioId) metadata.set("x-scenario-id", args.scenarioId);
  const scenarioSuffix = args.scenarioId ? ` | scenario=${args.scenarioId}` : "";
  const staggerSuffix = args.staggerMs ? ` | stagger=${args.staggerMs}ms` : "";

  const csv: ResultCSV | null = args.csv ? new ResultCSV(args.csv) : null;
  if (csv) console.error(`# csv=${args.csv}`);

  console.log(
    `📞 Calling ${args.target} | source=${path.basename(audioPath)} ` +
      `(${wavDurationS(wav).toFixed(2)}s @ ${wav.rate}Hz/${wav.channels}ch ${wav.wireFormat}) ` +
      `| duration=${args.duration.toFixed(1)}s | frame=${args.chunkMs}ms | ` +
      `concurrency=${args.concurrency}${staggerSuffix}${scenarioSuffix} ` +
      `| transport=${transportLabel("client")}`,
  );
  console.log("─".repeat(70));

  const activeCalls: Call[] = [];
  const shutdown = installSignalShutdown(activeCalls);

  const client = new DeepfakeDetectionClient(args.target, channelCredentials());
  let results: CallResult[];
  try {
    results = await Promise.all(
      labels.map((label, i) =>
        runOneCall(
          client, label, wav, args.chunkMs, args.duration,
          metadata, i * args.staggerMs, csv, path.basename(audioPath),
          (c) => activeCalls.push(c),
        ),
      ),
    );
  } finally {
    client.close();
    if (csv) await csv.close();
  }

  const failures = results.filter((r) => r.error !== null);
  if (shutdown.seen) {
    process.stderr.write("─".repeat(70) + "\n");
    process.stderr.write(
      `Shutdown complete (${args.concurrency - failures.length}/${args.concurrency} streams finished cleanly).\n`,
    );
    return;
  }
  console.log("─".repeat(70));
  const ok = args.concurrency - failures.length;
  console.log(`Summary: ${ok}/${args.concurrency} streams OK, ${failures.length} failed`);
  for (const { label, error } of failures) {
    const e = error as ServiceError | Error;
    const code = (e as ServiceError).code !== undefined ? `code=${(e as ServiceError).code} ` : "";
    console.error(`  [${label}] ${code}${e.message ?? String(e)}`);
  }
  if (failures.length > 0) process.exit(1);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/python/phone_call_burst.py
"""N concurrent live phone calls → DetectDeepfake.

The recommended multi-call integration pattern: when a single PBX (e.g.
a FreeSWITCH instance running `mod_audio_fork`) is handling many calls
simultaneously, you want **one** long-lived gRPC channel multiplexing
N concurrent bidi streams — not N separate channels. This file shows
exactly that shape:

  - One `make_aio_channel(target)` for the whole run.
  - N concurrent `stub.DetectDeepfake()` bidi streams over that channel,
    each running the same `send_call` / `recv_call` loop that
    `phone_call.py` documents in detail (imported, not duplicated).
  - Per-stream `call-NN` label prefixes every log line so interleaved
    output is grep-friendly.
  - Optional `--stagger-ms` to spread call starts out (mimics a real
    PBX where calls arrive over time, not all at exactly t=0).
  - Graceful SIGINT/SIGTERM via `common.install_signal_shutdown` cancels
    every in-flight stream cleanly — no orphan bidi sockets, summary
    still prints.
  - Optional `--csv PATH` captures per-chunk results across all streams
    in one file — useful for finding the connection-count knee on a
    real backend (where do errors start? when does median latency
    degrade? when does the GPU saturate?).

CLI:
    python phone_call_burst.py [--audio FILE] [--target localhost:50051]
                               [--chunk-ms 100] [--duration 30]
                               [--concurrency 5] [--stagger-ms 500]
                               [--scenario-id ID] [--csv PATH]

Defaults:
  - --concurrency 1 (effectively `phone_call.py`)
  - --stagger-ms 0   (thundering-herd: every call starts at the same instant)
  - --audio omitted  → picks the first .wav in `examples/audio/`
"""

from __future__ import annotations

import argparse
import asyncio
import sys
import time
from pathlib import Path

import grpc

from aurigin.deepfake_detection.v1 import deepfake_detection_pb2_grpc as pb_grpc

from common import (
    ResultCSV, WavData,
    install_signal_shutdown, make_aio_channel, read_wav, transport_label,
)
# The per-call building blocks. `phone_call_burst` is "send_call + recv_call
# instantiated N times over one channel" — importing keeps that relationship
# explicit and prevents drift between the two files' send/recv loops.
from phone_call import recv_call, send_call

DEFAULT_CHUNK_MS = 100
DEFAULT_DURATION_S = 30.0
DEFAULT_CONCURRENCY = 1


def _resolve_audio(arg: Path | None) -> Path:
    if arg is not None:
        if not arg.is_file():
            raise FileNotFoundError(arg)
        return arg
    audio_dir = Path(__file__).resolve().parent.parent / "audio"
    wavs = sorted(audio_dir.glob("*.wav")) if audio_dir.is_dir() else []
    if not wavs:
        raise SystemExit(
            "No --audio supplied and no .wav files found in examples/audio/. "
            "Drop a 16-bit PCM or 32-bit IEEE-float WAV in examples/audio/ or pass one with --audio.",
        )
    return wavs[0]


async def _run_one_call(
    stub, label: str, wav: WavData, chunk_ms: int, duration_s: float,
    metadata: tuple[tuple[str, str], ...], delay_s: float,
    csv_out: ResultCSV | None, file_name: str,
) -> tuple[str, BaseException | None]:
    """Open one bidi stream, run send_call + recv_call concurrently. Returns
    (label, exception_or_none) so a failing call doesn't take the others
    down — multi-call survival counts need to be meaningful.

    The per-call work is exactly what phone_call.py does (same imports,
    same loops); the multi-call coordination — fan-out, stagger, signal
    handler, summary — lives in main()."""
    if delay_s > 0:
        await asyncio.sleep(delay_s)
    call = stub.DetectDeepfake(metadata=metadata) if metadata else stub.DetectDeepfake()
    sink: dict = {
        "session_id": "", "chunks": [],
        "audio_duration_ms": 0, "global_result": "unknown",
    }
    t_start = time.perf_counter()
    try:
        await asyncio.gather(
            send_call(call, wav, chunk_ms, duration_s),
            recv_call(call, label=label, sink=sink),
        )
        result: BaseException | None = None
    except BaseException as exc:  # CancelledError, AioRpcError, etc.
        result = exc
    finally:
        if csv_out is not None:
            csv_out.write_session(
                file_name, sink["session_id"], sink["chunks"],
                sink["audio_duration_ms"], sink["global_result"],
                (time.perf_counter() - t_start) * 1000.0,
            )
    return label, result


def _make_labels(n: int) -> list[str]:
    width = max(2, len(str(n)))
    return [f"call-{i:0{width}d}" for i in range(1, n + 1)]


async def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__.split("\n", maxsplit=1)[0])
    parser.add_argument("--audio", type=Path, default=None, help="WAV file to stream (defaults to first in audio/)")
    parser.add_argument("--duration", type=float, default=DEFAULT_DURATION_S, help="Per-call length in seconds")
    parser.add_argument("--chunk-ms", type=int, default=DEFAULT_CHUNK_MS, help="Audio frame size in milliseconds")
    parser.add_argument("--target", default="localhost:50051", help="gRPC server host:port")
    parser.add_argument(
        "-c", "--concurrency", type=int, default=DEFAULT_CONCURRENCY,
        help="Number of concurrent streams to open over a single channel.",
    )
    parser.add_argument(
        "--stagger-ms", type=int, default=0,
        help="Delay between successive concurrent kickoffs in ms. 0 = all calls start at the same instant.",
    )
    parser.add_argument(
        "--scenario-id", default=None,
        help="Server-side simulator scenario to request (sent as x-scenario-id metadata).",
    )
    parser.add_argument("--csv", default=None, help="Write per-chunk results to this path (overwrites)")
    args = parser.parse_args()

    if args.concurrency < 1:
        parser.error("--concurrency must be >= 1")
    if args.stagger_ms < 0:
        parser.error("--stagger-ms must be >= 0")

    audio_path = _resolve_audio(args.audio)
    wav = read_wav(audio_path)
    labels = _make_labels(args.concurrency)
    metadata: tuple[tuple[str, str], ...] = (
        (("x-scenario-id", args.scenario_id),) if args.scenario_id else ()
    )
    scenario_suffix = f" | scenario={args.scenario_id}" if args.scenario_id else ""
    stagger_suffix = f" | stagger={args.stagger_ms}ms" if args.stagger_ms else ""

    csv_out: ResultCSV | None = ResultCSV(args.csv) if args.csv else None
    if csv_out is not None:
        print(f"# csv={args.csv}")

    print(
        f"📞 Calling {args.target} | source={audio_path.name} "
        f"({wav.duration_s:.2f}s @ {wav.rate}Hz/{wav.channels}ch {wav.wire_format}) "
        f"| duration={args.duration:.1f}s | frame={args.chunk_ms}ms | "
        f"concurrency={args.concurrency}{stagger_suffix}{scenario_suffix} "
        f"| transport={transport_label()}",
    )
    print("─" * 70)

    try:
        async with make_aio_channel(args.target) as channel:
            stub = pb_grpc.DeepfakeDetectionStub(channel)

            stagger_s = args.stagger_ms / 1000
            tasks = [
                asyncio.create_task(_run_one_call(
                    stub, label, wav, args.chunk_ms, args.duration,
                    metadata, delay_s=i * stagger_s,
                    csv_out=csv_out, file_name=audio_path.name,
                ))
                for i, label in enumerate(labels)
            ]
            # Graceful Ctrl-C: cancel each per-call task. _run_one_call's
            # `except BaseException` converts CancelledError into a clean
            # (label, exc) tuple so asyncio.gather still completes with all
            # results — we get the usual summary instead of a traceback.
            shutdown = install_signal_shutdown(tasks)
            results = await asyncio.gather(*tasks)
    finally:
        if csv_out is not None:
            csv_out.close()

    failures = [(label, exc) for label, exc in results if exc is not None]
    if shutdown.seen:
        print("─" * 70, file=sys.stderr)
        print(
            f"Shutdown complete ({args.concurrency - len(failures)}/{args.concurrency} streams finished cleanly).",
            file=sys.stderr,
        )
        return
    print("─" * 70)
    ok = args.concurrency - len(failures)
    print(f"Summary: {ok}/{args.concurrency} streams OK, {len(failures)} failed")
    for label, exc in failures:
        if isinstance(exc, grpc.aio.AioRpcError):
            print(f"  [{label}] gRPC error: {exc.code().name}: {exc.details()}", file=sys.stderr)
        else:
            print(f"  [{label}] {type(exc).__name__}: {exc}", file=sys.stderr)
    if failures:
        sys.exit(1)


def cli() -> None:
    """Sync entrypoint for `uv run phone-call-burst`."""
    loop = asyncio.new_event_loop()
    try:
        loop.run_until_complete(main())
    finally:
        loop.close()


if __name__ == "__main__":
    cli()
